geogram: init at 1.8.3

hesiod opened the pull request:
###### Description of changes

Add the [Geogram](https://github.com/BrunoLevy/geogram) library.
Part of an ongoing effort to package [Meshroom](https://github.com/alicevision/Meshroom) (https://github.com/NixOS/nixpkgs/issues/94127).
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

hesiod committed:
geogram: init at 1.8.3

davidak commented on pkgs/development/libraries/geogram/default.nix:
```diff
@@ -0,0 +1,91 @@
+  cmakeFlags = [
+    # Triangle has unclear licensing situation
+    "-DGEOGRAM_WITH_TRIANGLE=OFF"
+    "-DGEOGRAM_WITH_LEGACY_NUMERICS=OFF"
```
why are the 3 other `OFF`?

hesiod replied:
see above

davidak commented on pkgs/development/libraries/geogram/default.nix:
```diff
@@ -0,0 +1,91 @@
+  '';
+
+  # The multiple output hook will remove $out during fixup, so we need to create it again
+  # to prevent Nix from complaining it is missing
```
that sound like a general problem. it would be great to find a general solution

could you report it?

hesiod replied:
I will first need to make sure this is not because of my own fault :slightly_smiling_face:
Perhaps I misconfigured something, not sure. Parts of the multiple output machinery seem to operate in.. mysterious ways.

davidak replied:
I'm sure `$out` is needed for a build to finish successfully. I remember it being the case for package tests.

So removing it seem a strange behavior.

hesiod replied:
resolved in amended commit

davidak commented:
Result of `nixpkgs-review pr 215728` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>geogram</li>
  </ul>
</details>

davidak commented on pkgs/development/libraries/geogram/default.nix:
```diff
@@ -0,0 +1,91 @@
+  outputs = [ "bin" "lib" "dev" "out" ];
+
+  cmakeFlags = [
+    # Triangle has unclear licensing situation
```
it's very clearly "unfree" in our context

https://github.com/BrunoLevy/geogram/wiki/ThirdParty#geometry

> [triangle (optional)](https://www.cs.cmu.edu/~quake/triangle.html), license: "Although Triangle is freely available, it is copyrighted by the author and may not be sold or included in commercial products without a license" 2D constrained Delaunay tiangulation.

Can you make it optional and when enabled, add "unfree" license to the package?

Not sure if it's worth this effort tho.
```suggestion
    # Triangle has unfree license, see https://www.cs.cmu.edu/~quake/triangle.html
```

hesiod replied:
I think I'd just leave it for the moment, since I've also disabled a few other package options. I'm packaging GeoGram for Meshroom, and AFAIK Meshroom doesn't need Triangle.

Regarding the other options:
* If `GEOGRAM_WITH_LEGACY_NUMERICS` is enabled GeoGram will build its own version of ARPACK, CBLAS, CLAPACK, LIBF2C and SUPERLU, which seems wasteful. Everything still seems to work when turning it off, so I just did that.
* `GEOGRAM_WITH_LUA`: Same story, GeoGram will build its own Lua interpreter. Meshroom doesn't need it so I turned it off.
* `GEOGRAM_WITH_GRAPHICS`: Pulls in a vendored version GLFW (or a system version if another flag is set). Again, I'm not really sure what it's used for, but Meshroom works without it so I turned it off.

I figured it's a good idea to drop the compile time a little by disabling nonessential flags.

davidak replied:
Sounds reasonable. Can you add that in a comment in the code so it can be later understood?

hesiod replied:
Sure, will do

davidak reviewed (changes requested):
Builds successful. The result has only 2 files. Is that expected?

```
[nix-shell:~/.cache/nixpkgs-review/pr-215728]$ tree results/
results/
└── geogram -> /nix/store/j64i56kvgg5wp5m1snhxwmm9wcbg2zm1-geogram-1.8.2-bin

1 directory, 0 files

[nix-shell:~/.cache/nixpkgs-review/pr-215728]$ tree results/geogram/
results/geogram/
└── bin
    ├── geodump
    └── vorpalite

1 directory, 2 files
```

The binaries start without error.

Not sure how to review the other outputs.

Generally looks good. Please see the other comments.

hesiod commented on pkgs/development/libraries/geogram/default.nix:
```diff
@@ -0,0 +1,91 @@
+      Geogram contains the main results in Geometry Processing from the former ALICE Inria project,
+      that is, more than 30 research articles published in ACM SIGGRAPH, ACM Transactions on Graphics,
+      Symposium on Geometry Processing and Eurographics.
+      It was supported by two grants from the European Research Council (ERC): GOODSHAPE and VORPALINE.
```
Sure, I'll remove it. I simply the first paragraph from their README :slightly_smiling_face:

davidak replied:
Yes, i have noticed that.

hesiod commented:
> Builds successful. The result has only 2 files. Is that expected?

It's a multiple output package, the interesting part is in the `lib` output. The binaries aren't that important.

davidak commented:
Library is successfully built:

```
[nix-shell:~/.cache/nixpkgs-review/pr-215728-4]$ tree result-lib
result-lib
└── lib
    ├── libgeogram.so -> libgeogram.so.1
    ├── libgeogram.so.1 -> libgeogram.so.1.8.2
    └── libgeogram.so.1.8.2
```

hesiod commented:
Marked as broken on `aarch64-linux` (some upstream bug)
@davidak I addressed all your comments, the PR should be ready for another review (if you have the time)

hesiod commented:
Result of `nixpkgs-review pr 215728` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>geogram</li>
  </ul>
</details>

davidak commented on pkgs/development/libraries/geogram/default.nix:
```diff
@@ -0,0 +1,110 @@
+  pname = "geogram";
+  version = "1.8.3";
+
+  src = fetchurl {
```
is there a reason you did not use `fetchFromGitHub` here?

hesiod replied:
Geogram requires submodules, but they provide a custom tarball (not one of the autogenerated Github ones) containing the sources including all submodules. This is a bit faster than `fetchFromGitHub` with submodules enabled. I have no strong preferences either way.

davidak commented:
Result of `nixpkgs-review pr 215728` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>geogram</li>
    <li>geogram.bin</li>
    <li>geogram.dev</li>
    <li>geogram.doc</li>
    <li>geogram.lib</li>
  </ul>
</details>

davidak commented on pkgs/development/libraries/geogram/default.nix:
```diff
@@ -0,0 +1,110 @@
+    zlib
+  ];
+
+  patches = [
```
can you try to upstream the patches?

the bundled zlib could be optional

hesiod replied:
The zlib patch is probably not acceptable for upstream as it doesn't provide any fallback path if zlib isn't present. There are ways to provide that using CMake wizardry, but that would be quite a large deviation from the current patch.
Upstream is already looking into improving the bundled library situation: https://github.com/BrunoLevy/geogram/issues/49

The other patch might be upstreamable. I'll look into it.

hesiod replied:
See https://github.com/BrunoLevy/geogram/pull/76

hesiod replied:
I also included a link to the issue in a comment above the patch. I think this should be good for now.

davidak replied:
Perfect. Thank you!

davidak commented on pkgs/development/libraries/geogram/default.nix:
```diff
@@ -0,0 +1,110 @@
+    python3Packages.robotframework
+  ];
+
+  # There are test failures as of version 1.8.3
```
do you see a path to where we can enable them?

are they few enough that we could disable only the failing tests?
could we report them upstream, so they can fix them?
or are they caused by nix sandbox build?

hesiod replied:
The tests were succeeding as of 1.8.2, but suddenly fail in 1.8.3. However, I have no idea why they fail (there are no error messages) and I'm not really familiar with robotframework.
I'm assuming that the failures are due to upstream bugs (since they were not present in 1.8.2), but I can't say for sure.

davidak commented on pkgs/development/libraries/geogram/default.nix:
```diff
@@ -0,0 +1,110 @@
+    runHook postCheck
+  '';
+
+  meta = {
```
```suggestion
  meta = with lib; {
```

i think this is more elegant

hesiod replied:
done

davidak commented on pkgs/development/libraries/geogram/default.nix:
```diff
@@ -0,0 +1,110 @@
+    '';
+    homepage = "https://github.com/BrunoLevy/geogram";
+    license = lib.licenses.bsd3;
+    # Broken on aarch64-linux as of version 1.8.3
```
```suggestion
    # Broken on aarch64-linux as of version 1.8.3
    broken = stdenv.isLinux && stdenv.isAarch64;
```

this would be better, so the metadata contains that it's broken

can you create an upstream bug about the issue and link it?

hesiod replied:
done

davidak commented on pkgs/development/libraries/geogram/default.nix:
```diff
@@ -0,0 +1,110 @@
+    homepage = "https://github.com/BrunoLevy/geogram";
+    license = lib.licenses.bsd3;
+    # Broken on aarch64-linux as of version 1.8.3
+    platforms = [ "x86_64-linux" "x86_64-darwin" "aarch64-darwin" ];
```
```suggestion
    platforms = [ "x86_64-linux" "aarch64-linux" "x86_64-darwin" "aarch64-darwin" ];
```

hesiod replied:
done

davidak reviewed (changes requested):
generally looks good

check my comments

hesiod commented:
Currently working on a few improvements to the checkPhase so I marked the PR as a draft, other than that the PR is in a good shape already

hesiod commented:
I've enabled the tests again, although I had to disable a few failing tests. I've also improved the checkPhase a bit (previously it would not actually abort the build if some tests failed)

davidak commented:
Result of `nixpkgs-review pr 215728` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>geogram</li>
    <li>geogram.bin</li>
    <li>geogram.dev</li>
    <li>geogram.doc</li>
    <li>geogram.lib</li>
  </ul>
</details>

davidak reviewed (approved):
Code looks good

Builds, binaries run
